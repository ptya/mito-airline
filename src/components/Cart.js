import React, { useState, useContext } from "react";
import PropTypes from "prop-types";

import Modal from "./Modal";
import Flight from "./Flight";
import { CartContext } from "./CartProvider";
import { StationsContext } from "./StationsProvider";

import { useSticky } from "./hooks/useSticky";
import { formatMoney } from "../utils/formatMoney";

const Cart = props => {
  const [payToggle, setPayToggle] = useState(false);
  const { cart, cartDispatch } = useContext(CartContext);
  const { origin, destination } = useContext(StationsContext);

  const handleReset = () => {
    setPayToggle(false);
    cartDispatch({
      type: "purge"
    });
  };

  const { className } = props;

  const total = formatMoney(cart.total);

  const isSticky = useSticky(100);

  const blockClass = isSticky
    ? `${className} cart cart--fixed`.trim()
    : `${className} cart`.trim();

  return (
    <>
      <div className={blockClass}>
        <div className="cart__details">
          <h2 className="cart__header">
            Flights <span className="cart__header--price">{total}</span>
          </h2>
          {!cart.outbound && (
            <p className="cart__info">Choose an outbound flight</p>
          )}
          {cart.outbound && (
            <Flight flight={cart.outbound} from={origin} to={destination} />
          )}
          {cart.inbound && (
            <Flight
              className={cart.outbound ? "cart__separator" : ""}
              flight={cart.inbound}
              from={destination}
              to={origin}
            />
          )}
          <h2 className="cart__total">
            Total <span>{total}</span>
          </h2>
        </div>
        <button
          type="button"
          className={
            cart.outbound ? "cart__btn" : "cart__btn cart__btn--disabled"
          }
          disabled={cart.outbound ? false : true}
          onClick={() => setPayToggle(true)}
        >
          Pay Now
        </button>
      </div>
      {payToggle && (
        <Modal setToggle={setPayToggle}>
          <div className="pay-view">
            <h1 className="pay-view__header">
              Thanks for buying your tickets at mito airlines
            </h1>
            <div className="pay-view__flights">
              <Flight flight={cart.outbound} from={origin} to={destination} />
              {cart.inbound && (
                <Flight flight={cart.inbound} from={destination} to={origin} />
              )}
            </div>
            <div className="pay-view__footer">
              <h2 className="pay-view__total">
                TOTAL: <span className="pay-view__total--amount">{total}</span>
              </h2>
              <button
                className="pay-view__cancel"
                type="button"
                onClick={() => handleReset()}
              >
                No, thanks (reset)
              </button>
            </div>
          </div>
        </Modal>
      )}
    </>
  );
};

Cart.propTypes = {
  className: PropTypes.string
};

Cart.defaultProps = {
  className: ""
};

export default Cart;
