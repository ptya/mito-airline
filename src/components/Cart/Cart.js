import React, { useState, useContext } from "react";
import PropTypes from "prop-types";
import { useTransition, config } from "react-spring";

// components
import CartFlight from "./Flight/CartFlight";
import CartModal from "./Modal/CartModal";

// context
import { CartContext } from "components/providers/CartProvider";
import { StationsContext } from "components/providers/StationsProvider";

// hooks
import { useSticky } from "components/hooks/useSticky";

// local elements
import Price from "./elements/Price";
import AnimatedWrapper from "./elements/AnimatedWrapper";
import AnimatedFlight from "./elements/AnimatedFlight";

// local styles
import CartWrapper from "./styles/CartWrapper";
import Header from "./styles/Header";
import Info from "./styles/Info";
import PayBtn from "./styles/PayBtn";
import Total from "./styles/Total";

const Cart = props => {
  const [payToggle, setPayToggle] = useState(false);

  const {
    cart: { total, outbound, inbound }
  } = useContext(CartContext);

  const {
    stations: { origin, destination }
  } = useContext(StationsContext);

  const { area } = props;

  const isSticky = useSticky(100);

  // modal animation
  const transition = useTransition(payToggle, null, {
    from: { opacity: 0, transform: "translate3d(0, -40px, 0)" },
    enter: { opacity: 1, transform: "translate3d(0,0,0)" },
    leave: { opacity: 0, transform: "translate3d(0,-40px,0)" },
    config: { config: config.stiff }
  });

  return (
    <>
      <CartWrapper area={area} isSticky={isSticky}>
        <Header>
          Flights <Price price={total} />
        </Header>
        <AnimatedWrapper>
          {!outbound && <Info>Choose an outbound flight</Info>}
          {outbound && (
            <AnimatedFlight>
              <CartFlight
                isSeparated={false}
                flight={outbound}
                from={origin}
                to={destination}
              />
            </AnimatedFlight>
          )}
          {inbound && (
            <AnimatedFlight>
              <CartFlight
                isSeparated={outbound ? true : false}
                flight={inbound}
                from={destination}
                to={origin}
              />
            </AnimatedFlight>
          )}
        </AnimatedWrapper>
        <Total>
          Total <Price price={total} />
        </Total>
        <PayBtn
          type="button"
          disabled={outbound ? false : true}
          onClick={() => setPayToggle(true)}
        >
          Pay Now
        </PayBtn>
      </CartWrapper>
      {transition.map(
        ({ item: toggle, key, props: animation }) =>
          toggle && (
            <CartModal
              key={key}
              setToggle={setPayToggle}
              animation={animation}
            />
          )
      )}
    </>
  );
};

Cart.propTypes = {
  area: PropTypes.string
};

Cart.defaultProps = {
  area: null
};

export default Cart;
