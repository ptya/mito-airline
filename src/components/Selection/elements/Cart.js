import styled from "styled-components";

const Cart = styled.div`
  grid-area: cart;
`;

export default Cart;
