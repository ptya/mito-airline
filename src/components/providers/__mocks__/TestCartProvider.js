import React, { useReducer } from "react";
import PropTypes from "prop-types";

import { CartContext, cartReducer, initialCart } from "../CartProvider";

const CartProvider = props => {
  const { state, reducer } = props;
  const [cart, cartDispatch] = useReducer(reducer, state);

  return (
    <CartContext.Provider
      value={{
        cart,
        cartDispatch
      }}
    >
      {props.children}
    </CartContext.Provider>
  );
};

CartProvider.propTypes = {
  children: PropTypes.node.isRequired,
  state: PropTypes.object,
  reducer: PropTypes.func
};

CartProvider.defaultProps = {
  state: initialCart,
  reducer: cartReducer
};

export default CartProvider;
