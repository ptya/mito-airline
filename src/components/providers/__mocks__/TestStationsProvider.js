import React, { useReducer } from "react";
import PropTypes from "prop-types";

import {
  StationsContext,
  stationsReducer,
  initialStations
} from "../StationsProvider";

const TestStationsProvider = props => {
  const { state, reducer } = props;
  const [stations, stationsDispatch] = useReducer(reducer, state);

  return (
    <StationsContext.Provider
      value={{
        stations,
        stationsDispatch
      }}
    >
      {props.children}
    </StationsContext.Provider>
  );
};

TestStationsProvider.propTypes = {
  children: PropTypes.node.isRequired,
  state: PropTypes.object,
  reducer: PropTypes.func
};

TestStationsProvider.defaultProps = {
  state: initialStations,
  reducer: stationsReducer
};

export default TestStationsProvider;
