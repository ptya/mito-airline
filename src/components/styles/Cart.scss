.cart {
  width: 230px;
  color: #343434;
}

.cart--fixed {
  position: fixed;
  top: 85px;
}

.cart__details {
  background: #FFF;
  border-top: 1px solid #c1c1c1;
}
.cart__header {
  padding: 15px 15px 0 15px;
  display: flex;
  justify-content: space-between;
  margin: 0;
  font-weight: bold;
  font-size: 14px;
  text-transform: uppercase;
}

.cart__header--price {
  color: #06038D;
}

.cart__info {
  padding: 15px;
  background: #FFF;
  font-family: Source Sans Pro;
  font-weight: bold;
  font-size: 16px;
  color: #919191;
  margin: 0;
}

.cart__separator {
  border-top: 1px dotted #C1C1C1;

  &:before {
    content: "";
    position: absolute;
    top: -7px;
    right: 0;
    left: 0;
    margin: auto;
    border-top: 10px solid white;
    border-left: 10px solid #ffffff;
    transform: rotate(45deg);
    box-shadow: 1px 1px #c1c1c1;
    width: 0;
  }
}

.cart__total {
  font-family: "Source Sans Pro";
  padding: 15px;
  display: flex;
  justify-content: space-between;
  background: #06038D;
  margin: 0;
  font-weight: bold;
  text-transform: uppercase;
  font-size: 16px;
  color: #FFF;
}


.cart__btn {
  background: #C6007E;
  border-radius: 3px;
  margin: 20px auto;
  width: 100%;
  height: 50px;
  border: 0;
  color: #fff;
  font-family: "Roboto";
  font-weight: 900;
  font-size: 16px;
  text-transform: uppercase;
}

.cart__btn--disabled {
  background: #c1c1c1;
}

.pay-view {
  min-width: 465px;
}
.pay-view__header {
  font-weight: 300;
  font-size: 16px;
  line-height: normal;
  text-transform: uppercase;
  color: #06038D;
  background: #F2F2F2;
  margin: 0;
  padding: 15px 30px;
}
.pay-view__flights {
  display: flex;
  padding: 15px 15px 0 15px;
  justify-content: space-between;
}
.pay-view__footer {
  display: flex;
  justify-content: space-between;
  margin-bottom: 10px;
}
.pay-view__total {
  padding: 15px 30px;
  margin: 0;
  font-weight: 500;
  font-size: 20px;
  text-transform: uppercase;
  color: #343434;
}

.pay-view__total--amount {
  font-weight: bold;
  color: #06038D;
}

.pay-view__cancel {
  background: #FFF;
  border: 0;
  padding: 5px 15px;
  margin: 10px 15px;
  font-weight: bold;
  font-size: 12px;
  text-decoration-line: underline;
  text-transform: uppercase;
  color: #C6007E;
  cursor: pointer;
}