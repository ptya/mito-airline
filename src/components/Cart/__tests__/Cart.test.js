import React from "react";
import {
  render,
  fireEvent,
  waitForElementToBeRemoved
} from "@testing-library/react";

// components
import Cart from "../Cart";

// context
import TestCartProvider from "components/providers/__mocks__/TestCartProvider";
import TestStationsProvider from "components/providers/__mocks__/TestStationsProvider";

// mocks
import {
  stationOne,
  stationTwo,
  flight
} from "components/providers/__mocks__/mockValues";

afterEach(() => {
  console.error.mockClear();
});

console.error = jest.fn();

function renderCart(cartState) {
  const stationsState = {
    origin: stationOne,
    destination: stationTwo
  };
  return render(
    <TestStationsProvider state={stationsState}>
      <TestCartProvider state={cartState}>
        <Cart />
      </TestCartProvider>
    </TestStationsProvider>
  );
}

test("Renders without stations and matches snapshot", () => {
  const { container } = renderCart();
  expect(container).toMatchSnapshot();
  expect(console.error).not.toHaveBeenCalled();
});

test("Renders with outbound flight and matches snapshot", () => {
  const { container } = renderCart({
    inbound: null,
    outbound: flight,
    total: 1000
  });
  expect(container).toMatchSnapshot();
  expect(console.error).not.toHaveBeenCalled();
});

test("Renders with inbound flight and matches snapshot", () => {
  const { container } = renderCart({
    inbound: flight,
    outbound: null,
    total: 100
  });
  expect(container).toMatchSnapshot();
  expect(console.error).not.toHaveBeenCalled();
});

test("Renders with all flights and matches snapshot", () => {
  const { container } = renderCart({
    inbound: flight,
    outbound: flight,
    total: 100
  });
  expect(container).toMatchSnapshot();
  expect(console.error).not.toHaveBeenCalled();
});

test("Clicking pay now toggles modal", () => {
  const { getByTestId, queryByTestId } = renderCart({
    inbound: flight,
    outbound: flight,
    total: 100
  });

  expect(queryByTestId("cm-btn")).toBeFalsy();
  fireEvent.click(getByTestId("pay-btn"));
  expect(getByTestId("cm-btn")).toBeTruthy();
});

test("In modal, clicking no thanks, resets cart", async () => {
  const { getByTestId, queryByTestId, getAllByText, getByText } = renderCart({
    inbound: flight,
    outbound: flight,
    total: 100
  });

  expect(queryByTestId("cm-btn")).toBeFalsy();

  fireEvent.click(getByTestId("pay-btn"));
  expect(getByTestId("cm-btn")).toBeTruthy();

  fireEvent.click(getByTestId("cm-btn"));
  await waitForElementToBeRemoved(() => getByTestId("my-portal"));

  expect(queryByTestId("cm-btn")).toBeFalsy();
  expect(getAllByText("$0")).toBeTruthy();
  expect(getByText("Choose an outbound flight")).toBeTruthy();
  expect(getByTestId("pay-btn").disabled).toBeTruthy();
});
