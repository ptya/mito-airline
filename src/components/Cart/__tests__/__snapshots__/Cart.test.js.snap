// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders with all flights and matches snapshot 1`] = `
<div>
  <aside
    class="sc-VigVT ebyAXx"
  >
    <h2
      class="sc-jTzLTM hRaYVe"
    >
      Flights 
      <span>
        $1
      </span>
    </h2>
    <div
      class="sc-gZMcBi gzvzg"
      style="height: 0px; overflow: hidden;"
    >
      <div>
        <div
          class="sc-gqjmRU VqbEQ"
          style="transform: translateY(-38px) rotateX(-90deg);"
        >
          <div
            class="sc-bdVaJa jmlIJf"
            data-testid="cart-flight"
          >
            <p
              class="sc-bwzfXH cWAdWB"
            >
              <span>
                Jul
              </span>
              <span>
                12
              </span>
            </p>
            <p
              class="sc-htpNat hPUCVW"
            >
              <span>
                Barcelona
              </span>
              <span>
                Budapest
              </span>
              <span>
                Fri
                 
                10:30
                 – 
                13:30
              </span>
            </p>
          </div>
        </div>
        <div
          class="sc-gqjmRU VqbEQ"
          style="transform: translateY(-38px) rotateX(-90deg);"
        >
          <div
            class="sc-bdVaJa eKdzhd"
            data-testid="cart-flight"
          >
            <p
              class="sc-bwzfXH cWAdWB"
            >
              <span>
                Jul
              </span>
              <span>
                12
              </span>
            </p>
            <p
              class="sc-htpNat hPUCVW"
            >
              <span>
                Budapest
              </span>
              <span>
                Barcelona
              </span>
              <span>
                Fri
                 
                10:30
                 – 
                13:30
              </span>
            </p>
          </div>
        </div>
      </div>
    </div>
    <h2
      class="sc-jTzLTM sc-cSHVUG guEnyY"
      data-testid="c-total"
    >
      Total 
      <span>
        $1
      </span>
    </h2>
    <button
      class="sc-jzJRlG bHvPj"
      data-testid="pay-btn"
      type="button"
    >
      Pay Now
    </button>
  </aside>
</div>
`;

exports[`Renders with inbound flight and matches snapshot 1`] = `
<div>
  <aside
    class="sc-VigVT ebyAXx"
  >
    <h2
      class="sc-jTzLTM hRaYVe"
    >
      Flights 
      <span>
        $1
      </span>
    </h2>
    <div
      class="sc-gZMcBi gzvzg"
      style="height: 0px; overflow: hidden;"
    >
      <div>
        <p
          class="sc-fjdhpX fUOfry"
        >
          Choose an outbound flight
        </p>
        <div
          class="sc-gqjmRU VqbEQ"
          style="transform: translateY(-38px) rotateX(-90deg);"
        >
          <div
            class="sc-bdVaJa jmlIJf"
            data-testid="cart-flight"
          >
            <p
              class="sc-bwzfXH cWAdWB"
            >
              <span>
                Jul
              </span>
              <span>
                12
              </span>
            </p>
            <p
              class="sc-htpNat hPUCVW"
            >
              <span>
                Budapest
              </span>
              <span>
                Barcelona
              </span>
              <span>
                Fri
                 
                10:30
                 – 
                13:30
              </span>
            </p>
          </div>
        </div>
      </div>
    </div>
    <h2
      class="sc-jTzLTM sc-cSHVUG guEnyY"
      data-testid="c-total"
    >
      Total 
      <span>
        $1
      </span>
    </h2>
    <button
      class="sc-jzJRlG iJfsJF"
      data-testid="pay-btn"
      disabled=""
      type="button"
    >
      Pay Now
    </button>
  </aside>
</div>
`;

exports[`Renders with outbound flight and matches snapshot 1`] = `
<div>
  <aside
    class="sc-VigVT ebyAXx"
  >
    <h2
      class="sc-jTzLTM hRaYVe"
    >
      Flights 
      <span>
        $10
      </span>
    </h2>
    <div
      class="sc-gZMcBi gzvzg"
      style="height: 0px; overflow: hidden;"
    >
      <div>
        <div
          class="sc-gqjmRU VqbEQ"
          style="transform: translateY(-38px) rotateX(-90deg);"
        >
          <div
            class="sc-bdVaJa jmlIJf"
            data-testid="cart-flight"
          >
            <p
              class="sc-bwzfXH cWAdWB"
            >
              <span>
                Jul
              </span>
              <span>
                12
              </span>
            </p>
            <p
              class="sc-htpNat hPUCVW"
            >
              <span>
                Barcelona
              </span>
              <span>
                Budapest
              </span>
              <span>
                Fri
                 
                10:30
                 – 
                13:30
              </span>
            </p>
          </div>
        </div>
      </div>
    </div>
    <h2
      class="sc-jTzLTM sc-cSHVUG guEnyY"
      data-testid="c-total"
    >
      Total 
      <span>
        $10
      </span>
    </h2>
    <button
      class="sc-jzJRlG bHvPj"
      data-testid="pay-btn"
      type="button"
    >
      Pay Now
    </button>
  </aside>
</div>
`;

exports[`Renders without stations and matches snapshot 1`] = `
<div>
  <aside
    class="sc-VigVT ebyAXx"
  >
    <h2
      class="sc-jTzLTM hRaYVe"
    >
      Flights 
      <span>
        $0
      </span>
    </h2>
    <div
      class="sc-gZMcBi gzvzg"
      style="height: 0px; overflow: hidden;"
    >
      <div>
        <p
          class="sc-fjdhpX fUOfry"
        >
          Choose an outbound flight
        </p>
      </div>
    </div>
    <h2
      class="sc-jTzLTM sc-cSHVUG guEnyY"
      data-testid="c-total"
    >
      Total 
      <span>
        $0
      </span>
    </h2>
    <button
      class="sc-jzJRlG iJfsJF"
      data-testid="pay-btn"
      disabled=""
      type="button"
    >
      Pay Now
    </button>
  </aside>
</div>
`;
